File: leetcode/70-climbing-stairs/solution.cpp
class Solution {

  int m[50];

public:
    int climbStairs(int n) {
      if (n == 1 || n == 2) {
        return n;
      }
        
      while (n>=0) {
        int res = m[n];

        if (res != NULL) {
          return res;
        }

        res = climbStairs(n -1) + climbStairs(n -2);

        m[n] = res;
      }

      return 0;
    }
};

File: leetcode/70-climbing-stairs/solution.java
import java.util.HashMap;
import java.util.Map;

class Solution {
    private static final Map<Integer, Integer> m = new HashMap<>();

    public int climbStairs(int n) {
        if (n == 1) return 1;
        if (n == 2) return 2;

        while (n >=0) {
            var res = m.get(n);

            if (res != null) return res;

            res = climbStairs(n-1) + climbStairs(n-2);
            
            m.put(n, res);
        }

        return 0;
    }
}
